lint-rules.md

```diff
@@ -0,0 +1,1266 @@
+        Int ncpu = 8
+        Int modify_disk_size_gb = 0
+    }
+```
+
+### `disallowed_input_name`
+
+Any input name matching these regular expressions will be flagged: `/^[iI]n[A-Z_]/` or `/^input/i`. It is redundant and needlessly verbose to use an input's name to specify that it is an input. Input names should be short yet descriptive. Prefixing a name with `in` or `input` adds length to the name without adding clarity or context.
+
+**Group**: `naming`
+
+#### Example
+
+All the below input names would be flagged:
+
+```wdl
+    input {
+        File inBam  # This name will be flagged twice, once for the 'in' prefix, and once for being camalCase when it should be snake_case
+        File in_bam
+        File input_gtf
+    }
+```
+
+### `disallowed_output_name`
+
+Any output name matching these regular expressions will be flagged: `/^[oO]ut[A-Z_]/`, `/^output/i` or `/^..?$/`. It is redundant and needlessly verbose to use an output's name to specify that it is an output. Output names should be short yet descriptive. Prefixing a name with `out` or `output` adds length to the name without adding clarity or context. Additionally, names with only 2 characters can lead to confusion and obfuscates the content of an output. Output names should be at least 3 characters long.
+
+**Group**: `naming`
+
+#### Example
+
+All the below input names would be flagged:
+
+```wdl
+    input {
+        File outBam  # This name will be flagged twice, once for the 'out' prefix, and once for being camalCase when it should be snake_case
+        File out_bam
+        File output_gtf
+        String af
+    }
+```
+
+### `no_curly_commands`
+
+`command` blocks should be wrapped with arrows (`<<<` `>>>`) instead of curly brackets (`{` `}`). Certain Bash constructions cause problems with parsing the bracket notation. There are no such problems with the arrow notation.
+
+#### Example
+
+Good:
+
+```wdl
+    command <<<
+        echo "Hello, World!"
+    >>>
+```
+
+Bad:
+
+```wdl
+    command {
+        echo "Hello, World!"
+    }
+```
+
+### `inpersistent_container`  && `persistent_container_not_tagged`
+
+All tasks should run in a persistent container. This ensures reproducibility across time and environments. `wdl-grammar` and `wdl-ast` will look for an `sha sum` tag in your `container` declarations and warn if one is missing. An `sha` digest will always point to the exact same image, whereas tags are mutable. This mutability makes even versioned tags problematic when we want to ensure reproducibility.
+
+While the confidence in persistence gained by using `sha` digests for pulling container images is valuable, it comes at a cost. Which is a lack of human readability. So we enforce preceeding `container` entries with a comment which gives a human readable name to the image being pulled. It is imperative that the `sha` digest and the human readable tag are kept in sync. It's extra overhead compared to just using a tag directly, but we consider it well worth it.
```

Comment:
I was unaware of this syntax at time of writing, and I would say this is the ideal syntax to enforce. _BUT_, Andrew T looked into it and found out that `singularity` doesn't support this syntax. So we can't use it internally.

I think this has to be one of the rules where we offer alternatives. However, I opted to leave all alternatives out of this first draft. Mainly due to the great length without going over variations. I didn't want to prolong the document.

My vote is update this rule to use the syntax proposed by @adthrasher . It seems ideal in all regards to me. However we can't use it, for now at least, so when it comes to implement this rule in code, we'll need to add some alternatives available to toggle. But the above stays the default, us as users would need to configure to something else.

Reply:
This issue (https://github.com/containers/buildah/issues/1407) provides some discussion of the `:tag@digest` format. Particularly this comment

```
Earlier versions of Docker, including 1.13.1, interpret this as "ask for the image with the specified tag, then fail if the manifest does not match the specified digest". That's fairly brittle but ultimately plausible (where the tag ~documents the intent and the digest ensure immutability and prevents moving the tag maliciously).

Later versions of Docker interpret this as "ask for the image with the specified digest, ignore the tag completely", which is, in my view, mostly pointless and can be actively misleading. If the tag is ignored in the code, the users' intent can just as well be noted in a comment after the command instead of inside the image specification, and that would be more accurate WRT what is actually happening.
```

Assuming that is accurate and docker is ignoring the `tag`, then I would tend to agree with the author that putting the `tag` in a comment is a better solution (and it aligns with your format).

Reply:
The "earlier" interpretation is what I assumed, so thank you for proving that assumption wrong. I would say the "later" interpretation is misleading and I'm surprised that's what Docker moved to. I think this needs some more investigation+discussion before settling on what the "most sane" default is.

I'll elaborate on what I intend to accomplish with this rule then pose a question to all the reviewers.

IMO WDL should strive to be reproducible across time and environments. The same WDL code should produce the same results regardless of where and when it was run. With this understanding, "moving" tags like `latest` should be avoided.

For context, our current `main` for `stjudecloud/workflows` uses BioContainer release tags (not `sha` digests) OR pulls `ghcr.io` images we own. To my understanding, the BioContainer tags are static and are never updated in place (rendering an `sha` digest redundant). We are in control over whether our `ghcr.io` images are overwritten, and we don't do that (though our CI _would_ allow it if we deviated from our standard practice). I think this is a perfectly acceptable set up.

The issue is that while writing a linter, the linter can't know the pulled tags are static or "safe". Which is why I argue the default for the linter should include an `sha` digest. However this also poses a problem, in that plain `sha` digests don't convey what version of a tool/image is being pulled. Hence the documents rule of also specifying a human readable version in a comment.

I understand that this rule, requiring `sha` digests in pulled container images, might be the most controversial rule in the document. So I wonder if we should list out what we consider "acceptable" alternatives for this rule. I'm hesitant to list alternate configurations for this rule, because it might obligate us (me) to list alternatives to other rules. I'd rather not do that, as the word count on this document is high enough as it is.

So, does this topic have enough nuance+import that it warrants a set of alternative configurations? Should any other rules get alternatives as well?

Reply:
I will go on record as saying I'm not a fan of this rule. I think there are very valid reasons to use non-SHASUM tags as long as you are aware of what you are doing. I also think the specifics of this rule make it hard to reasonable about. But, if you two are of the opinion we should keep it, then let's keep it.